### PR TITLE
36 value backed storage

### PR DESCRIPTION
**Pull request by konistehrad**
XXX: failing tests, not ready for merge. See comments here: https://github.com/thesecretlab/YarnSpinner/issues/36#issuecomment-202226300

### CODE REVIEW

File: YarnSpinner/Dialogue.cs [hunk 1]
```diff
@@ -330,54 +349,27 @@ private class StandardLibrary : Library {
...
-						return parameters[0].AsBool == parameters[1].AsBool;
-					case Value.Type.Null:
-						// Only null-null comparisons are true.
-						return parameters[0].type == Value.Type.Null;
-					}
-
-					// Give up and say they're not equal
-					return false;
-
+                    return parameters[0].Equals( parameters[1] );
```
Comment (desplesda) [inline YarnSpinner/Dialogue.cs]: So much nicer. :+1:

File: YarnSpinner/Value.cs [hunk 2]
```diff
@@ -0,0 +1,334 @@
+
+using System;
+
+namespace Yarn
+{
+	// A value from inside Yarn.
+	public class Value : IComparable, IComparable<Value> {
```
Comment (desplesda) [inline YarnSpinner/Value.cs]: Breaking this out into its own file is a good idea.

Comment (konistehrad) [inline YarnSpinner/Value.cs]: Once the overloaded operators started showing up it was almost mandatory. Commenting all the JS semantics take up a ton of space, haha.

File: YarnSpinner/Value.cs [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+			Bool,    // a boolean value
+			Variable, // the name of a variable; will be expanded at runtime
+			Null,    // the null value
+		}
+
+		public Value.Type type { get; internal set; }
+        
+        // this is primarily a debugging property, allowing
+        // for testing between actual and perceived type
+        internal object rawBacking { get; set; }
```
Comment (desplesda) [inline YarnSpinner/Value.cs]: Do we really need to have this? It means that we're storing data twice, and that makes me nervous. 

If we need a stronger way to ensure that the type that the Value thinks it is is the same as the value that was applied, adding logic to the underlying values (`numberValue`, `stringValue`, etc) that throws an exception if the wrong underlying value was accessed (eg if `numberValue` is accessed, but the Value is a bool, then throw an exception) might be a better idea. What do you think?

File: YarnSpinnerTests/Test.cs [hunk 4]
```diff
@@ -101,13 +102,26 @@ public void TestIndentation()
...
+
+			var path = System.IO.Path.Combine ("TestCases", "VariableStorage.node");
+			dialogue.LoadFile (path);
+
+			foreach (var result in dialogue.Run()) {
+				HandleResult (result);
+			}
+		}
+
+		// [Test()]
```
Comment (desplesda) [inline YarnSpinnerTests/Test.cs]: A lot of the  `Test` attributes have been commented out - were you encountering errors in the unit tests?

Comment (konistehrad) [inline YarnSpinnerTests/Test.cs]: Nope, sorry was just reducing the cases surrounding my own test. Was gonna restore them when the final PR dropped. :)

File: YarnSpinner/Value.cs (hunk 3 above)
Comment (konistehrad) [inline YarnSpinner/Value.cs]: I'll come clean here: I added that exclusively as a debug variable initially (to ensure the input to the constructor resulted in good state thruout the rest of the class), then made the mistake of leveraging against it in `Equals` and `GetHashCode`. I'll redo the latter and pull the `rawBacking` prop.

**Comment by konistehrad** [comment]
Okay! I updated this, fixed some bad comparisons, updated my test cases, restored the others, and everything looks green across the board! Let me know what you think! :)

**Comment by desplesda** [comment]
Noticing that the Travis builds are failing - you can run the tests locally without having to wait for travis by doing this on the terminal:

```
nuget restore YarnSpinner.sln
xbuild /p:Configuration=Release YarnSpinner.sln
mono ./YarnSpinnerTests/bin/Release/YarnSpinnerTests.exe
./run-tests.sh -d Tests/TestCases
````

**Comment by konistehrad** [comment]
Just saw that @desplesda sorry. One of those days; tests are passing, you push the code, then tests fail. Egg on my face. 😓

**Comment by desplesda** [comment]
Haha, no sweat! Happens to the best of us.

**Comment by konistehrad** [comment]
Welp! This one is out of my pay grade: this will exit with 0 on local, but is clearly exiting with 1 on CI. No leads on this, sorry!!

**Comment by desplesda** [comment]
Ok, will have a chance to look at this hopefully soon! 👌

**Comment by konistehrad** [comment]
It was a failing test case in Options.node, sorry I'm on it.

**Comment by konistehrad** [comment]
Adding Type.Undefined and using it introduced some nastiness, despite my best intentions. I checked against Twine, and using an undefined variable, as the initial implementation suggests, should be Type.Null not Type.Undefined. I've pulled references to Type.Undefined and made Type.Null the default.

Extending that, I've made `assert` a variadic function in Twine, allowing a description of the failing test case. While not line numbers or the like (I saw that TODO in there), it does allow for pinpointing the failure state.

Sorry that my onboarding has been a little rocky. I promise I'll get the hang of this test harness. 😉

**Comment by desplesda** [comment]
Looking good! I'll play with this a little before merging its but it all looks solid. 

Please bug me if I haven't either merged this or spotted a problem after a few days!

File: YarnSpinner/Value.cs [hunk 5]
```diff
@@ -0,0 +1,351 @@
+
+using System;
+
+namespace Yarn
+{
+	// A value from inside Yarn.
+	public class Value : IComparable, IComparable<Value> {
+		public static readonly Value DEFAULT = new Value();
```
Comment (desplesda) [inline YarnSpinner/Value.cs]: This feels a bit weird. It looks like the `DEFAULT` value is only used for the case where a variable cannot be found, and is always null; if that's the case, why isn't it called null? And if that change was made, shouldn't all null values simply be references to this instance?

Comment (konistehrad) [inline YarnSpinner/Value.cs]: @desplesda Love it, that fix is in! I wasn't able to remove the default constructor, however, as it's still needed to create Type.Variable Value's, as seen in `Parser.cs:694`. Any guidance you could provide would be much appreciated.

**Comment by desplesda** [comment]
Awesome work.

**Pull request merged**